megatron/tokenizer/gpt2_tokenization.py

```diff
@@ -230,20 +230,81 @@ def bpe(self, token):
+        # matter since the English models were not trained on any Chinese data
+        # and generally don't have any Chinese data in them (there are Chinese
+        # characters in the vocabulary because Wikipedia does have some Chinese
+        # words in the English Wikipedia.).
+        output = []
+        for char in text:
+            cp = ord(char)
+            if self._is_chinese_char(cp):
+                output.append(" ")
+                output.append(char)
+                output.append(" ")
+            else:
+                output.append(char)
+        return "".join(output)
+      
+    #copied from https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/tokenization_bert.py
+    def _is_chinese_char(self, cp):
+        """Checks whether CP is the codepoint of a CJK character."""
+        # This defines a "chinese character" as anything in the CJK Unicode block:
+        #   https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_(Unicode_block)
+        #
+        # Note that the CJK Unicode block is NOT all Japanese and Korean characters,
+        # despite its name. The modern Korean Hangul alphabet is a different block,
+        # as is Japanese Hiragana and Katakana. Those alphabets are used to write
+        # space-separated words, so they are not treated specially and handled
+        # like the all of the other languages.
+        if (
+            (cp >= 0x4E00 and cp <= 0x9FFF)
+            or (cp >= 0x3400 and cp <= 0x4DBF)  #
+            or (cp >= 0x20000 and cp <= 0x2A6DF)  #
+            or (cp >= 0x2A700 and cp <= 0x2B73F)  #
+            or (cp >= 0x2B740 and cp <= 0x2B81F)  #
+            or (cp >= 0x2B820 and cp <= 0x2CEAF)  #
+            or (cp >= 0xF900 and cp <= 0xFAFF)
+            or (cp >= 0x2F800 and cp <= 0x2FA1F)  #
+        ):  #
+            return True
+
+        return False
+      
+    @lru_cache(1000000)
+    def normalize_token_py3(self, token):
+          token = ''.join(self.byte_encoder[b] for b in token.encode('utf-8'))
+          ret = [bpe_token for bpe_token in self.bpe(token).split(' ')]
+          return ret
+
+    def tokenize(self, text, max_token_len_cache=9, tokenize_chinese_chars=False):
         """ Tokenize a string. """
         bpe_tokens = []
+        if sys.version_info[0] == 2:
+          if tokenize_chinese_chars:
+            text = self._tokenize_chinese_chars(text)
+          for token in re.findall(self.pat, text):        
+              token = ''.join(self.byte_encoder[ord(b)] for b in token)
+              bpe_tokens.extend(bpe_token for bpe_token in self.bpe(token).split(' '))
+          return bpe_tokens
         for token in re.findall(self.pat, text):
-            if sys.version_info[0] == 2:
-                token = ''.join(self.byte_encoder[ord(b)] for b in token)
-            else:
+           if tokenize_chinese_chars:
+              text = self._tokenize_chinese_chars(text)
+           if len(token) <= max_token_len_cache:
```

Comment:
Hum so you don't want to cache tokens that are longer than 10 characters? Would you be able to benchmark that heuristic?

Reply:
See comment below. We benchmarked current implementatin of 9 char at 60 mb/s. I think upping this limit will make things marginally faster, but will take more memory. Everything is still capped at 1M normalized tokens.

Reply:
Okay as discussed offline, let's keep that.

Reply:
```suggestion
           if len(token) <= self.max_token_len_cache:
```

Reply:
```suggestion
          if len(token) <= max_token_len_cache:
```

indentation is very inconsistent in Meg. at the very least let's not introduce new indentation of 3.